### ISSUE TITLE
reportUnnecessaryComparison: False positive for attributes that get reassigned in another method

### ISSUE THREAD

**Issue by arvidfm** [microsoft/pyright#7011]
**Describe the bug**
If you perform a comparison on an instance attribute in e.g. a `while` loop, Pyright will narrow its type in the loop body even if the attribute can be reassigned elsewhere, leading to a reportUnnecessaryComparison false positive if you perform the same comparison later in the loop

Encountered in practice in asynchronous code (the loop might be cancelled while yielding to another coroutine).

**Code or Screenshots**
In this example the instance attribute gets reassigned in `do_stuff`, so it might be `None` after the call:

```python
class A:
  def __init__(self) -> None:
    self.value: int | None = 10

  def loop(self) -> None:
    while self.value is not None:
      self.do_stuff()

      # the attribute can get reassigned since
      # the while condition was checked
      if self.value is not None:
        self.send_status(self.value)

  def do_stuff(self) -> None:
    if self.value is None:
      return
    self.value -= 1
    if self.value < 5:
      # we reassign the attribute here
      self.value = None

  def send_status(self, value: int) -> None:
    pass
```

https://pyright-play.net/?code=MQAgDgngTglg5gCwC4C4QGcmwMZIFB7YA2AhuuiAIIp4ggAmApgGYgD6bMAdjEhwBTpGRZgEoQAWgB8IAHIB7Loxp06QkQDoAbiSIBXZSG5IQAHzmLGIALwgAjAAYCdJqyLz5YQcLGSZCpRVVAHcEGCIrdWZtXQMjCi55EwDlWlU1Hw16eTZMPWZmflFndKNWKJj9KxgEpItAtNKMTKEuelykEiQ9dG9NHSritNcGHLyCvt9petT0mHLMgbiamaD0qEZuqC5G5v7YqwlbO135veilqwAeEABWNdUKy5sZkpHW9swunsmAGhBLmhjOJpikHmAyOg8EA

**VS Code extension or command-line**
pyright 1.1.347 using the installer from PyPI

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

Type narrowing techniques used in static type checkers like pyright and mypy assume that the code doesn't change the type of a value within other execution contexts. This is generally a good assumption in well-written code, and it allows for better type narrowing and the reduction of false positives. Your code sample above violates this assumption. You'll need to make modifications to your code if you want it to work with static type checking.

**Comment by arvidfm** [comment]
I imagine it's a pretty common pattern in async code to do variations of

```py
while self._flag is not None:
   ...
```

and then have a `cancel()` method that sets `self._flag` to `None`. I'm not sure I agree that's necessarily code smell.

Incidentally, if I use a boolean instead of a nullable flag value Pyright doesn't complain about the repeat check which seems a bit inconsistent. Though from some testing it only seems to complain when type narrowing is involved, so maybe that's by design. E.g. it doesn't complain about this:

```py
from typing import Literal

n: Literal[True] = True
if n:
  pass
```

Either way, IMO Pyright seems to be jumping the gun a bit here when type narrowing instance attributes. Personally I'd rather be forced to write `while (flag := self._flag) is not None:` if I want to be able to rely on the type narrowing inside the body of the loop. Or at least have the *option* to be forced to write that. This is how smart casting works in Kotlin, for instance.

**Comment by erictraut** [comment]
Pyright's behavior is consistent here with other Python static type checkers (like mypy) and static type checkers in other languages (like TypeScript). We're not likely to change this behavior.
